packages/smart-wallet/src/smartWallet.js

```diff
@@ -1019,6 +1049,23 @@ export const prepareSmartWallet = (baggage, shared) => {
           await E(seatRef).tryExit();
         },
       },
+
+      invoke: {
+        /**
+         * @param {string} expr
+         */
+        async evalExpr(expr) {
+          trace('TODO: validate Justin syntax');
+          trace('evalExpr', expr);
+          // XXX worth using a nameHub vs. a mapStore?
+          const { nameHub, nameAdmin } = this.state.my;
+          const endowments = { E, harden, assert, nameHub, nameAdmin };
+          const c = new Compartment(endowments);
```

Comment:
`endowments` are additions to the HardenedJS default allowed properties of `c.globalThis`.  To restrict the Compartment's `globalThis` so that only `endowments` can be reached, I think you'd need something more like (untested):

```suggestion
          const c = new Compartment();
          for (const key of Reflect.ownKeys(c.globalThis)) {
            assert(Reflect.deleteProperty(c.globalThis, key));
          }
          for (const [key, value] of Object.entries(endowments)) {
            assert(Reflect.defineProperty(c.globalThis, key, { value }));
          }
          harden(c.globalThis);
```

Reply:
I'm not sure about the scope of your suggestion... could you give an example of an expression that shows how things work differently with the suggested approach?

Reply:
Given evaluation of a Justin expression:
```js
try {
  const ret = c.evaluate(`Number.isSafeInteger(123)`);
  console.log('@@@ returned', ret);
} catch (e) {
  console.log('@@@ caught', e);
}
```

With the existing code, all of HardenedJS's tamed globals, in addition to the explicit endowments, are reachable by `c.evaluate`:

```
@@@ returned true
```

With (something like) my proposed code, only the endowments are reachable, not any other globals, so the above `c.evaluate(...)` will instead do:

```
@@@ caught ReferenceError: Number is not defined
```

I suggest you indicate in a comment which of these behaviours you desire for the `c.evaluate`.

Reply:
It seems like `Number.isSafeInteger` should be available. Is there anything in the default environment of `Compartment` that carries authority?

I suppose compute time is relevant too... `[...].sort()` will be `O(n log(n))` rather than the usual case of `O(n)` for Justin exprs. I'm not sure denying that is practical, though. hm.

Reply:
> anything in... `Compartment` that carries authority?

Not in the sense of the ability to break integrity, short of what is added to `c.globalThis` (such as the properties of `additionalGlobals` in `new Compartment(additionalGlobals)`).

> I'm not sure denying that is practical

Just spitballing some ideas, neither of which is necessarily practical or desirable:

- Further restrict an individual Compartment's capabilities with an allowlist or membrane, but both of those require HardenedJS features that we don't have yet.
- Have the host vat spawn separate child vats to perform evaluations, but that doesn't scale well in the current SwingSet.

Reply:
Talking with @erights, it seems there's a sweet spot I hadn't considered:

- Have a multitenant vat provide a hand-written Justin interpreter:
  - The interpreter can take a method allowlist as a parameter, and also have restricted implementations of JS builtins
  - Interpreter-level metering can abort a particular evaluation long before an XS-level meter overflow would kill the vat

Reply:
@michaelfig , I agree with the first bullet. 

On the second bullet, yes, with care.

Normally we allow the Justin program to cause synchronous side effects. But if you terminate it with any meter-like mechanism and you don't kill the vat it is affecting, then you leave behind inconsistent corrupt state. Unless you
- limit it to only performing synchronous invocations that it could have performed remotely and asynchronously.
- start the Justin execution in its own turn.

Then, any of the methods it could invoke are designed to transition the vat from a consistent state to a consistent state by itself, without making assumptions about future incoming messages causing other effects.

IOW, sudden termination of such a local Justin computation is equivalent to a remote invoker who stops.